### PR TITLE
feat: `avm/res/app/container-app`

### PR DESCRIPTION
**Pull request by oZakari**
## Description

Closes avm/res/app/container-app [#128](https://github.com/Azure/Azure-Verified-Modules/issues/128#issuecomment-1828902446)

If you haven't already, read the full [contribution guide](https://github.com/Azure/bicep-registry-modules/blob/main/CONTRIBUTING.md). The guide may have changed since the last time you read it, so please double-check. Once you are done and ready to submit your PR, edit the PR description and run through the relevant checklist below.

Enable GitHub Workflows in your fork to enable auto-generation of assets with our [GitHub Action](/.github/workflows/push-auto-generate.yml).
To trigger GitHub Actions after auto-generation, [add a GitHub PAT](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token) as a secret in your forked repository called `PAT`.

> edit: @AlexanderSehr 
> | Pipeline |
> | - |
> | [avm.res.app.container-app](https://github.com/oZakari/bicep-registry-modules/actions/workflows/avm.res.app.container-app.yml) |

## Adding a new module

- [x] A proposal has been submitted and approved.
- [x] I have included "Closes #{module_proposal_issue_number}" in the PR description.
- [ ] I have run `brm validate` locally to verify the module files.
- [x] I have run deployment tests locally to ensure the module is deployable.

### CODE REVIEW

**Comment by segraef** [comment]
Please add a status check badge of your latest successful as proof. More details see [here](https://azure.github.io/Azure-Verified-Modules/contributing/bicep/bicep-contribution-flow/).

**Review by AlexanderSehr — CHANGES_REQUESTED** [review]
cc: @eriqua (I was so free to steal you comment from another PR)

> Hi @oZakari, codeowners file should be updated. For guidelines on how to configure teams please refer to https://azure.github.io/Azure-Verified-Modules/specs/shared/#team-memberships

This refers to the `.github/CODEOWNERS` file. You should be able to find a corresponding line and can uncomment it :)

File: avm/res/app/container-app/README.md [hunk 1]
```diff
@@ -0,0 +1,1047 @@
+# Container Apps `[Microsoft.App/containerApps]`
+
+This module deploys a Container App.
+
+## Navigation
```
Comment (AlexanderSehr) [inline avm/res/app/container-app/README.md:5]: Please update the version of your fork as there have been some updates to the ReadMe utility (like a disclaimer for data collection being added).

File: avm/res/app/container-app/main.bicep [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+output resourceId string = containerApp.id
+
+@description('The name of the resource group the Container App was deployed into.')
+output resourceGroupName string = resourceGroup().name
+
+@description('The name of the Container App.')
+output name string = containerApp.name
+
+@description('The principal ID of the system assigned identity.')
+output systemAssignedMIPrincipalId string = (managedIdentities.?systemAssigned ?? false) && contains(containerApp.identity, 'principalId') ? containerApp.identity.principalId : ''
```
Comment (AlexanderSehr) [inline avm/res/app/container-app/main.bicep:256]:
```suggestion
output systemAssignedMIPrincipalId string = containerApp.?identity.?principalId ?? ''
```

File: avm/res/app/container-app/main.bicep [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+@description('Optional. The diagnostic settings of the service.')
+param diagnosticSettings diagnosticSettingType
+
+@description('Optional. The managed identity definition for this resource.')
+param managedIdentities managedIdentitiesType
+
+@description('Optional. Array of role assignment objects that contain the \'roleDefinitionIdOrName\' and \'principalId\' to define RBAC role assignments on this resource. In the roleDefinitionIdOrName attribute.')
+param roleAssignments roleAssignmentType
+
+@description('Optional. Enable telemetry via a Globally Unique Identifier (GUID).')
```
Comment (AlexanderSehr) [inline avm/res/app/container-app/main.bicep:66]:
```suggestion
@description('Optional. Enable/Disable usage telemetry for module.')
```

File: avm/res/app/container-app/main.bicep [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+  @description('Optional. The name of logs that will be streamed. "allLogs" includes all possible logs for the resource. Set to \'\' to disable log collection.')
+  logCategoriesAndGroups: {
+    @description('Optional. Name of a Diagnostic Log category for a resource type this setting is applied to. Set the specific logs to collect here.')
+    category: string?
+
+    @description('Optional. Name of a Diagnostic Log category group for a resource type this setting is applied to. Set to \'AllLogs\' to collect all logs.')
+    categoryGroup: string?
+  }[]?
+
+  @description('Optional. The name of logs that will be streamed. "allLogs" includes all possible logs for the resource. Set to \'\' to disable log collection.')
```
Comment (AlexanderSehr) [inline avm/res/app/container-app/main.bicep:278]:
```suggestion
  @description('Optional. The name of metrics that will be streamed. "allMetrics" includes all possible metrics for the resource. Set to \'\' to disable metric collection.')
```

File: avm/res/app/container-app/version.json [hunk 5]
```diff
@@ -0,0 +1,7 @@
+{
+    "$schema": "https://aka.ms/bicep-registry-module-version-file-schema#",
+    "version": "0.4",
```
Comment (AlexanderSehr) [inline avm/res/app/container-app/version.json:3]:
```suggestion
    "version": "0.1",
```
Let's start low ;)

File: avm/res/batch/batch-account/main.json [hunk 6]
```diff
@@ -1486,4 +1486,4 @@
       "value": "[if(and(coalesce(tryGet(parameters('managedIdentities'), 'systemAssigned'), false()), contains(reference('batchAccount', '2022-06-01', 'full').identity, 'principalId')), reference('batchAccount', '2022-06-01', 'full').identity.principalId, '')]"
     }
   }
-}
\ No newline at end of file
+}
```
Comment (AlexanderSehr) [inline avm/res/batch/batch-account/main.json:1489]: Appears unintended. Please undo.

File: avm/res/app/container-app/tests/e2e/defaults/main.test.bicep [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+param resourceGroupName string = 'dep-${namePrefix}-app.containerApps-${serviceShort}-rg'
+
+@description('Optional. The location to deploy resources to.')
+param location string = deployment().location
+
+@description('Optional. A short identifier for the kind of deployment. Should be kept short to not run into resource-name length-constraints.')
+param serviceShort string = 'acamin'
+
+@description('Optional. Enable telemetry via a Globally Unique Identifier (GUID).')
+param enableTelemetry bool = true
```
Comment (AlexanderSehr) [inline avm/res/app/container-app/tests/e2e/defaults/main.test.bicep:21]:
```suggestion
```

File: avm/res/app/container-app/tests/e2e/defaults/main.test.bicep [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        eventHubAuthorizationRuleResourceId: diagnosticDependencies.outputs.eventHubAuthorizationRuleId
+        storageAccountResourceId: diagnosticDependencies.outputs.storageAccountResourceId
+        workspaceResourceId: diagnosticDependencies.outputs.logAnalyticsWorkspaceResourceId
+      }
+    ]
+    tags: {
+      'hidden-title': 'This is visible in the resource name'
+      Env: 'test'
+    }
+    enableTelemetry: enableTelemetry
```
Comment (AlexanderSehr) [inline avm/res/app/container-app/tests/e2e/defaults/main.test.bicep:82]:
```suggestion
```

In the min test, we should not have any parameters that are not absolutly necessary

File: avm/res/app/container-app/tests/e2e/max/main.test.bicep [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+param resourceGroupName string = 'dep-${namePrefix}-app.containerApps-${serviceShort}-rg'
+
+@description('Optional. The location to deploy resources to.')
+param location string = deployment().location
+
+@description('Optional. A short identifier for the kind of deployment. Should be kept short to not run into resource-name length-constraints.')
+param serviceShort string = 'acamax'
+
+@description('Optional. Enable telemetry via a Globally Unique Identifier (GUID).')
+param enableTelemetry bool = true
```
Comment (AlexanderSehr) [inline avm/res/app/container-app/tests/e2e/max/main.test.bicep:21]:
```suggestion
```

File: avm/res/app/container-app/tests/e2e/max/dependencies.bicep [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+resource managedIdentity 'Microsoft.ManagedIdentity/userAssignedIdentities@2023-01-31' = {
+  name: managedIdentityName
+  location: location
+}
+
+@description('The resource ID of the created Managed Identity.')
+output managedIdentityResourceId string = managedIdentity.id
+
```
[the remark this answers is not in the document]
Comment (AlexanderSehr) [reply avm/res/app/container-app/tests/e2e/max/dependencies.bicep:23]: For role assignment tests

File: avm/res/app/container-app/tests/e2e/waf-aligned/main.test.bicep [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+param resourceGroupName string = 'dep-${namePrefix}-app.containerApps-${serviceShort}-rg'
+
+@description('Optional. The location to deploy resources to.')
+param location string = deployment().location
+
+@description('Optional. A short identifier for the kind of deployment. Should be kept short to not run into resource-name length-constraints.')
+param serviceShort string = 'acawaf'
+
+@description('Optional. Enable telemetry via a Globally Unique Identifier (GUID).')
+param enableTelemetry bool = true
```
Comment (AlexanderSehr) [inline avm/res/app/container-app/tests/e2e/waf-aligned/main.test.bicep:21]:
```suggestion
```

File: avm/res/app/container-app/tests/e2e/waf-aligned/main.test.bicep [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+        eventHubAuthorizationRuleResourceId: diagnosticDependencies.outputs.eventHubAuthorizationRuleId
+        storageAccountResourceId: diagnosticDependencies.outputs.storageAccountResourceId
+        workspaceResourceId: diagnosticDependencies.outputs.logAnalyticsWorkspaceResourceId
+      }
+    ]
+    tags: {
+      'hidden-title': 'This is visible in the resource name'
+      Env: 'test'
+    }
+    enableTelemetry: enableTelemetry
```
Comment (AlexanderSehr) [inline avm/res/app/container-app/tests/e2e/waf-aligned/main.test.bicep:83]:
```suggestion
```

File: avm/res/batch/batch-account/main.json (hunk 6 above)
Comment (oZakari) [reply avm/res/batch/batch-account/main.json]: Thanks, have reverted this file back to upstream/main

**Comment by oZakari** [comment]
> cc: @eriqua (I was so free to steal you comment from another PR)
> 
> > Hi @oZakari, codeowners file should be updated. For guidelines on how to configure teams please refer to https://azure.github.io/Azure-Verified-Modules/specs/shared/#team-memberships
> 
> This refers to the `.github/CODEOWNERS` file. You should be able to find a corresponding line and can uncomment it :)

Thanks @AlexanderSehr, I have uncommented the line for this resource in the codeowners file.

File: avm/res/app/container-app/README.md (hunk 1 above)
Comment (oZakari) [reply avm/res/app/container-app/README.md:5]: I have updated my branch with upstream main and rerun Set-AVMModule.

File: avm/res/app/container-app/tests/e2e/waf-aligned/main.test.bicep [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+    location: location
+    lock: {
+      kind: 'CanNotDelete'
+      name: 'myCustomLockName'
+    }
+    managedIdentities: {
+      userAssignedResourceIds: [
+        nestedDependencies.outputs.managedIdentityResourceId
+      ]
+    }
```
[the remark this answers is not in the document]
Comment (oZakari) [reply avm/res/app/container-app/tests/e2e/waf-aligned/main.test.bicep:93]: MSI is required according to PSRule, However, secrets are not so I have pulled that out of the WAF-aligned test.

**Comment by oZakari** [comment]
Hey @AlexanderSehr, any idea what I am missing with this [test](https://github.com/oZakari/bicep-registry-modules/actions/runs/7590183096/job/20676250606#step:10:209)

Appears to be a couple of tests are failing:
- Param category check: I checked the regex for the roleAssigments param and it looks fine. image
- Telemetry identifier - The code looks correct from my perspective, so not sure what I am missing with this one.

**Comment by AlexanderSehr** [comment]
> Hey @AlexanderSehr, any idea what I am missing with this [test](https://github.com/oZakari/bicep-registry-modules/actions/runs/7590183096/job/20676250606#step:10:209)
> 
> Appears to be a couple of tests are failing:
> 
> * Param category check: I checked the regex for the roleAssigments param and it looks fine. image
> * Telemetry identifier - The code looks correct from my perspective, so not sure what I am missing with this one.

Hey @oZakari,

thanks for working through the comments. Regarding the failing tests I just checked your pipeline and would recommend to re-run it. The commit the pipeline uses is not the latest you actually pushed. 
As a result e.g. the `roleAssignement.condition` parameter in the pipeline run is actually missing a `.` at the end, which is already fixed in your latest version and hence should not lead to any failing test.

Regarding the telemetry, it is actually expecting a slightly different value, as you can see in the following:
- Expected: `46d3xbcp.res.app-containerapp` 
- Current: `46d3xbcp.res.app-container-app`
For reference: The value is taken from the CSV linked [here](https://azure.github.io/Azure-Verified-Modules/indexes/bicep/bicep-resource-modules/).
Again, I can see this was fixed in the latest commit, so please make sure the readme & json are the latest using the `Set-avmmodule -ModuleFolderPath '.\avm\res\app\container-app' -Recurse` utility and re-run the pipeline :)

File: avm/res/batch/batch-account/main.json [hunk 14]
```diff
@@ -1486,4 +1486,4 @@
       "value": "[if(and(coalesce(tryGet(parameters('managedIdentities'), 'systemAssigned'), false()), contains(reference('batchAccount', '2022-06-01', 'full').identity, 'principalId')), reference('batchAccount', '2022-06-01', 'full').identity.principalId, '')]"
     }
   }
-}
+}
```
Comment (AlexanderSehr) [inline avm/res/batch/batch-account/main.json:1489]: There seems to be an unintended change, likely done by auto-formatting.

File: .github/CODEOWNERS [hunk 15]
```diff
@@ -6,7 +6,7 @@
 #/avm/res/aad/domain-service/ @Azure/avm-res-aad-domainservice-module-owners-bicep
 /avm/res/analysis-services/server/ @Azure/avm-res-analysisservices-server-module-owners-bicep @Azure/avm-core-team-technical-bicep
 /avm/res/api-management/service/ @Azure/avm-res-apimanagement-service-module-owners-bicep @Azure/avm-core-team-technical-bicep
-#/avm/res/app/container-app/ @Azure/avm-res-app-containerapp-module-owners-bicep @Azure/avm-core-team-technical-bicep
+/avm/res/app/container-app/ @Azure/avm-res-app-containerapp-module-owners-bicep @Azure/avm-core-team-technical-bicep
```
Comment (AlexanderSehr) [inline .github/CODEOWNERS:9]: Please note, that the team must also be created as per: https://azure.github.io/Azure-Verified-Modules/specs/shared/#team-memberships (including the team's parent-child) relationship

File: avm/res/batch/batch-account/main.json (hunk 14 above)
Comment (oZakari) [reply avm/res/batch/batch-account/main.json:1489]: This had been removed in a prior commit, and doesn't look to exist in current PR anymore.

File: .github/CODEOWNERS (hunk 15 above)
Comment (oZakari) [reply .github/CODEOWNERS:9]: Just finished, thanks for the reminder!

**Comment by oZakari** [comment]
Hi @AlexanderSehr, I think I got everything sorted at this point. With your help of course, so thanks much! Let me know if you need anything else from me.

File: avm/res/app/container-app/main.bicep [hunk 16]
```diff
@@ -194,32 +191,6 @@ resource containerApp 'Microsoft.App/containerApps@2023-05-01' = {
   }
 }
 
-resource containerApp_diagnosticSettings 'Microsoft.Insights/diagnosticSettings@2021-05-01-preview' = [for (diagnosticSetting, index) in (diagnosticSettings ?? []): {
```
Comment (AlexanderSehr) [inline avm/res/app/container-app/main.bicep]: Hey @oZakari, how come you removed the diagnostic settings? Are they actually not supported on the resource?

Comment (AlexanderSehr) [reply avm/res/app/container-app/main.bicep]: If they are, please add it again - and also make sure it's part of the `max` & `waf-aligned` tests.

Comment (oZakari) [reply avm/res/app/container-app/main.bicep]: They are not supported at the container app unfortunately, only at the managed/app environment. See here for reference: https://github.com/microsoft/azure-container-apps/issues/382

**Comment by AlexanderSehr** [comment]
> Hi @AlexanderSehr, I think I got everything sorted at this point. With your help of course, so thanks much! Let me know if you need anything else from me.

Sad, but interesting. One would wish the RP would at least be so kind to throw an error. Meh

**Review by AlexanderSehr — APPROVED** [review]

File: .github/CODEOWNERS (hunk 15 above)
Comment (AlexanderSehr) [reply .github/CODEOWNERS:9]: Unresolving to merge only after the relationship was approved by the parent team's owner

Comment (eriqua) [reply .github/CODEOWNERS:9]: Hey @oZakari, as codeowners file shows errors, could you check these guidelines https://azure.github.io/Azure-Verified-Modules/specs/shared/#team-memberships ? It's mainly about creating relevant Azure teams and requesting them to be added to relevant parent teams. Let us know if you get stuck in the approval process so we can try to unblock it.

**Pull request merged**